_posts/2016-07-11-value-of-automated-e2e-testing.md

```diff
@@ -0,0 +1,58 @@
+
+The views in the user interface are designed in such a way that the user can see the main item and all the sub-items related at the same time. The main item could have several sub-items listed in a table-like manner. The main item has certain managing tools and actions and, of course, all the sub-items have a bunch of tools and actions too. Most of the tools are implemented as icons with actions attached. The state and visibility of the tools depend about the status of the main item, but also from the context. Is this the main item which created the sub-item, the main item which edited them or a main item having nothing to do with the others? In addition to, some notification icons need to be shown in certain situations. However, after all we had a number of different icons which should be visible if and only if the sunspots were in the right position. 
+
+At least for me, the variety of showing data depending on the context was also a burden. The functionality behind hiding and showing the icons respectively was very easy to break by sincere mistake when implementing some new feature or tweaking an existing one. Also, the number of different permutations were so vast that it was nearly impossible to make full regression testing just by clicking the web site manually. I feel this was the main reason we started to write automated end-to-end-tests in quite an early phase of the project.
+
+![A bunch of icons](/img/value-of-e2e-testing/ikonit.JPG)
+
+##The First Try with Wicket Unit Tests
+
+The first automated e2e-tests were written using [Wicket Unit Tests](https://cwiki.apache.org/confluence/display/WICKET/Unit+Test). The unit tests utilized WicketTester, which runs the application without the need to actually run a web server. You can validate the web page opening, click some buttons, fill fields, run the business logic and validate the web page after. The Wicket tests are pretty quick to run and fairly simple to write. But there are also disadvantages, the tests are not actually testing the full pipe, because the lack of web server and browser. Also, the Wicket page tester does not create a picture of the page, meaning that in a case of exception or assertion error, the actual web page cannot be seen. It is like trying to find a needle in a dark room. Furthermore, the actual page structure can be seen only in debug output, i.e. finding the components to assert is not always easy. 
+
+However, despite the impediments, these tests were found very helpful. The object was reached, the tests checked that the icons were visible and clickable in the specified situations, and even more, if you broke functionality by accident it was exposed by a test.
```

Comment:
"the object was reached" -> should this be objective? This sentence is also very long.